### ISSUE TITLE
Optimization deleting declared dependencies is wrong

### ISSUE THREAD

**Issue by cancerberoSgx** [gfranko/amdclean#83]
I noticed that your tool is removing dependencies if they are not referenced. This is wrong, take for example :

```javascript
define('SCA.Shopping.Starter', ['A','B','C'], function(A) {
	var theRestOfArgs = Array.prototype.slice.call(arguments, 1);
});
```

is transformed into: 

```javascript
SCAShoppingStarter = function (A) {
  var theRestOfArgs = Array.prototype.slice.call(arguments, 1);
}(A);
```

This is not equivalent to requirejs where the variable theRestOfArgs contains [B, C] and in amdclean transformed code will be empty.

You shouln't perform this optimization or perform it only if the arguments name is not used inj the callback.

Good tool!! congrats and thanks!

**Comment by gfranko** [comment]
Okay, I can look into making sure the params are passed.  Thanks for reporting this!

**Comment by cancerberoSgx** [comment]
I need this fix urgently so I hope I can send you a pull request soon. Thanks

**Comment by gfranko** [comment]
I can get a new release by Monday, if that works for you

**Comment by cancerberoSgx** [comment]
oh, by monday is more than fine for me, thanks! very appreciated. 

That's what i'm doing right now (listing all deps), but the thing is that I cannot release my code with that.

BTW, I think I've fixed the problem unfortunately I feel kind of dirty: in convertToFunctionExpression.js#255 y just added this fragment:

var callbackBody = amdclean.options.code.substring(callbackFunc.body.range[0], callbackFunc.body.range[1]); 
if(/[^\w0-9_]arguments[^\w0-9_]/.test(callbackBody))
{
  cbParams = dependencyNames;
}

This fix is working for me, but I'm afraid I've broken other use cases and certainly it breaks the unit tests :(

**Comment by gfranko** [comment]
Don't worry about it, I'll take care of it this weekend

**Comment by cancerberoSgx** [comment]
Thank you man, and congrats again for you code now that i've seen it!

**Comment by cancerberoSgx** [comment]
Hi, I'm sorry to put this in this issue, but I have to mention it.

I've noticed that in README you mention that the gain is only in file size. So after using amdclean I was waiting to see smaller files but the real gain for us was CPU time wasted on dependency calculation - more visible in big projects and small devices. 

I put some notes here http://themysteriesofjavascript.blogspot.com/2015/03/amdclean-not-documented-feature.html . 

I recommend you to mention that in your README. I can share some profiling charts comparing this if you want. Regards.

**Comment by gfranko** [comment]
@cancerberoSgx That's a really good point about eliminating the need for dependency resolution.  I would love to see your profiling charts.

Also, one of the reasons I originally built AMDclean was because I was writing a lot of JavaScript **libraries**, and wanted to use a module system without having to leave behind any trace of other third-party code (e.g. Browserify).  This is why the file size and readability is presented as a big pro.

**Comment by juansilva** [comment]
I was testing @cancerberoSgx solution and one thing I found was that the mix of declared arguments and omited argumets may results in issues, 
Take this code for instance:

define('a', function () {
	'use strict';
	return {};
});

define('b.tpl', function () {
	'use strict';
	return 'this is a test';
});

define('c', function () {
	'use strict';
	return {};
});


define('c', ['a', 'b.tpl', 'c'], function(a, b_tpl)
{
	'use strict';

	a;
	b_tpl;
	arguments;
});

"b_tpl" wil be replaced by "btpl", 
I modified it a bit and now it does keep named ones and adds the others:

      var callbackBody = callbackFunc.body ? amdclean.options.code.substring(callbackFunc.body.range[0], callbackFunc.body.range[1]) : '';
      if (/[^\w0-9_]arguments[^\w0-9_]/.test(callbackBody)) {
        cbParams = cbParams.concat(dependencyNames.slice(cbParams.length));
      }

Other thing I found with this solution is that arguments may not be refering to the modules arguments but to some inner function arguments, I didnot solve that one tho :(

BTW, This is awesome code @gfranko, thanks for it!!!

**Comment by cancerberoSgx** [comment]
I've just tested the fix from @juansilva and it works fine. Even unit tests pass OK. 
I will try to solve the 'false positive' when an inner function uses 'arguments'. It is not a big problem but it would be perfect. 
Will try to add unit tests for this issue and make a pull request since we need this issue to be solved urgently. Thanks.

**Comment by cancerberoSgx** [comment]
added fix for this issue in https://github.com/gfranko/amdclean/pull/84/files . checking for 'false positives' while referencing 'arguments' from inside inner functions. Unit tests made.

**Comment by cancerberoSgx** [comment]
Just to let you now that I've added some cpu profiling charts to my blog regarding gains in dependency resolution speed using big applications and slow devices: http://themysteriesofjavascript.blogspot.com/2015/03/amdclean-not-documented-feature.html .
BTW: do you have any news about this issue? Was the pull request good enough ?

**Comment by gfranko** [comment]
@cancerberoSgx I just merged your pull request and released **2.5.0**.  Also, would you mind sending a separate PR with updating the README with your performance findings about AMDclean?
